upcoming: [M3-7611] - Placement Groups Detail Summary

carrillo-erik opened the pull request:
## Description 📝
This PR introduces the `PlacementGroupsSummary` component.

## Changes  🔄
- Add Summary panel which displays the Placement Group's configuration details.
- Add logic to display a warning notice in the even the Placement Group is not compliant.
- Add unit testing for component.

## Preview 📷

### Placement Group Summary View
![pg-summary](https://github.com/linode/manager/assets/119514965/152fe64c-a191-4dcb-8b23-25259a56a0b8)

### Tooltip
![pg-summary-tooltip](https://github.com/linode/manager/assets/119514965/67d5041d-2a6e-430f-b7e6-415438f5982b)

### Non-compliant Placement Group Notice
![Screenshot 2024-02-09 at 6 09 41 AM](https://github.com/linode/manager/assets/119514965/b35e882a-6197-439c-8459-da51b8021029)

## How to test 🧪

### Prerequisites
(How to setup test environment)
- Enable the following using the Cloud Manager Dev Tools:
    - Placement Groups feature flag
    - MSW

### Verification steps 
(How to verify changes)
- Navigate to `http://localhost:3000/placement-groups`
- Click on a Placement Group from the table.
- Verify the configuration info to match the info from the Placement Group you clicked.
- Verify that the tooltip is rendered and behaves as expected.
- Run the following unit test: `yarn test PlacementGroupsSummary`

## As an Author I have considered 🤔
*Check all that apply*

- [ ] 👀 Doing a self review
- [ ] ❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
- [ ] 🤏 Splitting feature into small PRs
- [ ] ➕ Adding a changeset
- [ ] 🧪 Providing/Improving test coverage
- [ ] 🔐 Removing all sensitive information from the code and PR description
- [ ] 🚩 Using a feature flag to protect the release
- [ ] 👣 Providing comprehensive reproduction steps
- [ ] 📑 Providing or updating our documentation
- [ ] 🕛 Scheduling a pair reviewing session
- [ ] 📱 Providing mobile support
- [ ] ♿  Providing accessibility support

abailly-akamai reviewed (commented):
@carrillo-erik looks good - you need to:

- add a changeset
- fix the unit failing
- add a test for the PlacementGroupsSummary.tsx

also:
![Screenshot 2024-02-08 at 15 28 29](https://github.com/linode/manager/assets/130582365/534be03c-9c67-48ee-874d-5e2e8cd3098d)

- The spacing is off for the non-compliance banner. 
- The learn more link in the banner is probably not accessible

carrillo-erik commented:
> @carrillo-erik looks good - you need to:
> 
> * fix the unit failing
> * add a test for the PlacementGroupsSummary.tsx 
> * The spacing is off for the non-compliance banner.

* Fixed the unit test that was failing and added unit test for PlacementGroupsSummary.tsx
* I refactored the `Notice` and the spacing was addressed to reflect the mock up in better detail.

> * The learn more link in the banner is probably not accessible
* Do you mean as in the link/location is not ready yet?

abailly-akamai commented:
@carrillo-erik no, I wonder what we do for those dark blue links in dark mode (the "Learn more" in this case). Cause there is not enough contrast for it to be accessible and legible

carrillo-erik commented:
> @carrillo-erik no, I wonder what we do for those dark blue links in dark mode (the "Learn more" in this case). Cause there is not enough contrast for it to be accessible and legible

Oh sorry, I misunderstood your comment. Indeed, our a11y does suffer quite a bit in dark mode. One suggestion, for the links in particular, could be that we adopt the same guidelines as GitHub. Recently they made the change to show the underline for links by default. The contrast ratio would still require some work, however, I believe they would be more legible.

bnussman-akamai commented on packages/manager/src/features/PlacementGroups/PlacementGroupsDetail/PlacementGroupsSummary/PlacementGroupsSummary.test.tsx:
```diff
@@ -0,0 +1,41 @@
+
+import { PlacementGroupsSummary } from './PlacementGroupsSummary';
+
+const queryMocks = vi.hoisted(() => ({
+  usePlacementGroupQuery: vi.fn().mockReturnValue({}),
+}));
+
+vi.mock('src/queries/placementGroups', async () => {
+  const actual = await vi.importActual('src/queries/placementGroups');
+  return {
+    ...actual,
+    usePlacementGroupQuery: queryMocks.usePlacementGroupQuery,
+  };
+});
```
This code does not appear to be doing anything. 

```suggestion
```

carrillo-erik replied:
These code blocks have been removed.

abailly-akamai commented on packages/manager/src/features/PlacementGroups/PlacementGroupsDetail/PlacementGroupsSummary/PlacementGroupsSummary.tsx:
```diff
@@ -0,0 +1,79 @@
+
+  const regionLabel =
+    regions?.find((region) => region.id === placementGroup.region)?.label ??
+    placementGroup.region;
```
Not sure if defaulting to the PG region ID is needed - the PG is a required.

if there's an issue here, may mean something went wrong so maybe we could do an early return with `ErrorState` if `regionLabel` is undefined?

carrillo-erik replied:
As you mentioned, placement group is required here, no need for default value.

abailly-akamai commented on packages/manager/src/features/PlacementGroups/utils.ts:
```diff
@@ -32,3 +32,13 @@ export const affinityTypeOptions = Object.entries(AFFINITY_TYPES).map(
+    AFFINITY_TYPES[placementGroup?.affinity_type]
+  }) is non-compliant.
+We are working to resolve compliance issues so that you can continue assigning Linodes to this Placement Group. `;
+};
```
You can move this in `PlacementGroupSummary` as well. I don't think it will be reused

carrillo-erik replied:
Done.

abailly-akamai reviewed (commented):
Almost there. just a few fixes for cleanup and consistency

carrillo-erik commented on packages/manager/src/features/PlacementGroups/PlacementGroupsDetail/PlacementGroupsDetail.tsx:
```diff
@@ -108,7 +118,24 @@ export const PlacementGroupsDetail = () => {
+                    Learn more.
+                  </Link>
+                </Typography>
+              </Notice>
```
@abailly-akamai I had originally done this, however, having the `<Notice />` inside the `<PlacementGroupSummary />` was the reason why the unit test was failing. This led me to rethink about the component composition and I feel the `<Notice />` should live outside of the `<PlacementGroupSummary />` as this component is only meant to display the config details. It's a personal preference but if you believe otherwise I can make the change.

abailly-akamai replied:
I don't think that test failure should warrant this change. The notice only pertains to the summary so it should be there. You put it in the `SafeTabPanel` which isn't right. Let's put it back inside the summary and fix the test. I can help if you can't find the source of the failure

carrillo-erik replied:
Made the changes.

carrillo-erik replied:
I didn't use a constant and instead incorporated the string right in the component, this seems to have solved the issue I was encountering with the unit test.

carrillo-erik commented on packages/manager/src/features/PlacementGroups/PlacementGroupsDetail/PlacementGroupsDetail.tsx:
```diff
@@ -1,16 +1,20 @@
+import { useTheme } from '@mui/material';
 import * as React from 'react';
 import { useHistory, useParams } from 'react-router-dom';
+import { Link } from 'react-router-dom';
```
Done.

mjac0bs commented on packages/manager/src/features/PlacementGroups/PlacementGroupsDetail/PlacementGroupsSummary/PlacementGroupsSummary.tsx:
```diff
@@ -0,0 +1,103 @@
+              data-testid="pg-non-compliant-notice-link"
+              to={'#'}
+            >
+              Learn more.
+            </Link>
```
```suggestion
              Learn more
            </Link>.
```
Minor nitpick: we don't usually include the period in the link.

carrillo-erik replied:
Took care of this.

mjac0bs commented on packages/manager/src/features/PlacementGroups/constants.ts:
```diff
@@ -8,3 +8,5 @@ export const MAX_NUMBER_OF_LINODES_IN_PLACEMENT_GROUP_MESSAGE =
 export const PLACEMENT_GROUP_LINODES_ERROR_MESSAGE =
   'There was an error loading Linodes for this Placement Group.';
+
+export const PLACEMENT_GROUP_TOOLTIP_TEXT = `The Affinity Type and Region determine the maximum number of VMs per group`;
```
```suggestion
export const PLACEMENT_GROUP_TOOLTIP_TEXT = `The Affinity Type and Region determine the maximum number of VMs per group.`;
```
Nit: sentence needs a period.

mjac0bs replied:
Also: are we using VMs in copy, but referring to them as Linodes in code, or should "VMs" also be "Linodes" here?

carrillo-erik replied:
There's a lot of copy that is still being worked, I will reach out to our technical writer about this and make any necessary changes in the a separate PR.

mjac0bs commented on packages/manager/src/features/PlacementGroups/PlacementGroupsDetail/PlacementGroupsSummary/PlacementGroupsSummary.test.tsx:
```diff
@@ -0,0 +1,29 @@
+import { PlacementGroupsSummary } from './PlacementGroupsSummary';
+
+describe('PlacementGroups Summary', () => {
+  it('renders the placement group detail summary panel', () => {
```
Optional: we could also check if the tooltip exists.

carrillo-erik replied:
I included this check in the unit test.

mjac0bs reviewed (approved):
Looks good - summary page displays the expected data based on the mocked API  response. Tests are passing. Left a couple very minor suggestions.

Unrelated to this PR, I also noted that the text tooltip in dark mode appears to be the wrong link color. Might be worth its own follow up ticket.

![Screenshot 2024-02-09 at 2 05 29 PM](https://github.com/linode/manager/assets/114685994/285bbfec-79af-41a1-92ab-996300a23593)
